syntax = "proto3";

package resdb;

message Operation {
	enum CMD {
    NONE = 0;
    SET = 1;
    GET = 2;
    GETALLVALUES = 3;
    GETRANGE = 4;
    SET_WITH_VERSION = 5;
    GET_WITH_VERSION = 6;
    GET_ALL_ITEMS = 7;
    GET_KEY_RANGE = 8;
    GET_HISTORY = 9;
    GET_TOP = 10;
  }
	CMD cmd = 1;
	string key = 2;
	bytes value = 3;
}

message KVRequest {
	repeated Operation ops = 1;
	Operation.CMD cmd = 2;
	string key = 3;
	bytes value = 4;
  int32 version = 5;
  // For get key range
  string min_key = 6;
  string max_key = 7;
  // For get history for a key
  int32 min_version = 8;
  int32 max_version = 9;
  // For top history
  int32 top_number = 10;
}

message ValueInfo {
  bytes value = 2;
  int32 version = 3;
}

message Item {
  string key = 1;
  ValueInfo value_info = 2;
}

message Items {
  repeated Item item = 1;
}

message KVResponse {
    string key = 1;
    bytes value = 2;
    ValueInfo value_info = 3;
    Items items = 4;
}

